src/platform/windows/ipc/ipc_session.cpp

```diff
@@ -0,0 +1,267 @@
+        _frame_ready.store(true, std::memory_order_release);
+      }
+    }
+  }
+
+  void ipc_session_t::handle_secure_desktop_message(const std::vector<uint8_t> &msg) {
+    if (msg.size() == 1 && msg[0] == SECURE_DESKTOP_MSG) {
+      // secure desktop detected
+      BOOST_LOG(info) << "WGC can no longer capture the screen due to Secured Desktop, swapping to DXGI";
+      _should_swap_to_dxgi = true;
+    }
+  }
+
+
+  int ipc_session_t::init(const ::video::config_t &config, std::string_view display_name, ID3D11Device *device) {
+    _process_helper = std::make_unique<ProcessHandler>();
+    _config = config;
+    _display_name = display_name;
+    _device = device;
+    return 0;
+  }
+
+  void ipc_session_t::initialize_if_needed() {
+    if (_initialized.load()) {
+      return;
+    }
+
+    if (bool expected = false; !_initialized.compare_exchange_strong(expected, true)) {
+      // Another thread is already initializing, wait for it to complete
+      while (_initialized.load()) {
+        std::this_thread::sleep_for(std::chrono::milliseconds(10));
+      }
+      return;
+    }
+
+    // Check if properly initialized via init() first
+    if (!_process_helper) {
+      BOOST_LOG(debug) << "Cannot lazy_init without proper initialization";
+      return;
+    }
+
+    // Get the directory of the main executable
+    std::string exePathBuffer(MAX_PATH, '\0');
+    GetModuleFileNameA(nullptr, exePathBuffer.data(), MAX_PATH);
+    exePathBuffer.resize(strlen(exePathBuffer.data()));
+    std::filesystem::path mainExeDir = std::filesystem::path(exePathBuffer).parent_path();
+    std::filesystem::path exe_path = mainExeDir / "tools" / "sunshine_wgc_capture.exe";
+    if (!_process_helper->start(exe_path.wstring(), L"")) {
+      auto err = GetLastError();
+      BOOST_LOG(error) << "Failed to start sunshine_wgc_capture executable at: " << exe_path.wstring()
+                       << " (error code: " << err << ")";
+      return;
+    }
+
+    bool handle_received = false;
+
+    auto on_message = [this, &handle_received](const std::vector<uint8_t> &msg) {
+      handle_shared_handle_message(msg, handle_received);
+      handle_frame_notification(msg);
+      handle_secure_desktop_message(msg);
+    };
+
+    auto on_error = [](const std::string &err) {
+      BOOST_LOG(error) << "Pipe error: " << err.c_str();
+    };
+
+    auto on_broken_pipe = [this]() {
+      BOOST_LOG(warning) << "Broken pipe detected, forcing re-init";
+      _force_reinit.store(true);
+    };
+
+    auto anon_connector = std::make_unique<AnonymousPipeFactory>();
+
+    auto raw_pipe = anon_connector->create_server("SunshineWGCPipe");
+    if (!raw_pipe) {
+      BOOST_LOG(error) << "IPC pipe setup failed - aborting WGC session";
+      // No need to call cleanup() - RAII destructors will handle everything
+      return;
+    }
+    _pipe = std::make_unique<AsyncNamedPipe>(std::move(raw_pipe));
+
+    _pipe->wait_for_client_connection(3000);
```

Comment:
This fixed timeout seems brittle. It's possible something could happen (CPU spike, slow disk, whatever) that means the connection doesn't get established within 3 seconds. We also just continue here in that case and try to send data on a disconnected pipe.